# Dimensiones de calidad de datos

## 1. Exploración inicial

In [1545]:
""" 1.1 Importacion de librerias de ciencia de datos """
import pandas as pd
import numpy as np
import re

secop_df = pd.read_parquet('secop_bienes.parquet')

In [1546]:
""" 1.2 Dimensiones del dataset, Revisión de la cantidad de filas y columnas para dimensionar el volumen de datos."""
print('Filas:', secop_df.shape[0])
print('Columnas:', secop_df.shape[1])

Filas: 196391
Columnas: 36


#### Dimensiones del dataset
- El dataset cuenta con 196391 Filas y 36 Columnas

In [1547]:
""" 1.3 Obtencion de los nombres de las columnas. """
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)
secop_df.columns

Index(['id_contrato', 'nombre_entidad', 'departamento', 'ciudad', 'orden',
       'sector', 'rama', 'entidad_centralizada', 'tipo_de_contrato',
       'modalidad_de_contratacion', 'codigo_de_categoria_principal',
       'objeto_del_contrato', 'fecha_de_firma', 'fecha_de_inicio_del_contrato',
       'fecha_de_fin_del_contrato', 'duraci_n_del_contrato',
       'condiciones_de_entrega', 'proveedor_adjudicado', 'documento_proveedor',
       'es_pyme', 'es_grupo', 'g_nero_representante_legal',
       'nacionalidad_representante_legal', 'valor_del_contrato',
       'valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago',
       'dias_adicionados', 'habilita_pago_adelantado',
       'el_contrato_puede_ser_prorrogado', 'origen_de_los_recursos',
       'destino_gasto', 'estado_contrato', 'liquidaci_n',
       'obligaci_n_ambiental', 'urlproceso'],
      dtype='str')

In [1548]:
""" 1.4 Mostrar las primeras 5 filas del Dataset. """
secop_df.head()

,id_contrato,nombre_entidad,departamento,ciudad,orden,sector,rama,entidad_centralizada,tipo_de_contrato,modalidad_de_contratacion,codigo_de_categoria_principal,objeto_del_contrato,fecha_de_firma,fecha_de_inicio_del_contrato,fecha_de_fin_del_contrato,duraci_n_del_contrato,condiciones_de_entrega,proveedor_adjudicado,documento_proveedor,es_pyme,es_grupo,g_nero_representante_legal,nacionalidad_representante_legal,valor_del_contrato,valor_pagado,valor_facturado,valor_pendiente_de_pago,dias_adicionados,habilita_pago_adelantado,el_contrato_puede_ser_prorrogado,origen_de_los_recursos,destino_gasto,estado_contrato,liquidaci_n,obligaci_n_ambiental,urlproceso
0,CO1.PCCNTR.1000001,PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION TERRITORIAL AMAZONIA,Distrito Capital de Bogotá,No Definido,Territorial,Ambiente y Desarrollo Sostenible,Ejecutivo,Centralizada,Suministros,Mínima cuantía,V1.50161509,Contratar a monto agotable por el sistema de precios unitarios; sin formula de reajuste el suministro de elementos de aseo y limpieza e insumos de cafetería y restaurante; para los Parques Nacionales Naturales de Amacayacu; Cahuinarí; Río Puré y Yaigojé Apaporis,2019-06-19,2019-06-18,2019-12-31,No definido,Como acordado previamente,INDUSTRIAS GUERRERO Y COMPAÑIA S.A.S.,830130048,Si,No,No Definido,CO,17897916.0,17896237,17896237,1679,0,No Definido,Si,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.846428&isFromPublicArea=True&isModal=true&asPopupView=true'}
1,CO1.PCCNTR.1000507,AGENCIA LOGISTICA DE LAS FUERZAS MILITARES,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Suministros,Selección Abreviada de Menor Cuantía,V1.50131612,SUMINISTRO DE HUEVOS CON DESTINO A LOS COMEDORES DE TROPA DE LA BR-4 Y BR-14; ADMINISTRADOS POR LA AGENCIA LOGÍSTICA DE LAS FUERZAS MILITARES REGIONAL ANTIOQUIA CHOCO Y OTRAS POSIBLES UNIDADES QUE LO REQUIERAN,2019-06-23,2019-06-13,2019-11-29,Dia(s),Transporte incluido,Distribuidora Antioqueña AM,21788564,Si,No,Mujer,CO,390000000.0,389999685,389999685,315,0,No,No,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.841446&isFromPublicArea=True&isModal=true&asPopupView=true'}
2,CO1.PCCNTR.1000901,FUERZA AEROESPACIAL COLOMBIANA,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Mínima cuantía,V1.53102701,ADQUISICIÓN DE PRESILLAS BORDADAS AZULES EN CANUTILLO PARA OFICIALES DE INSIGNIA DE LA FUERZA AÉREA COLOMBIANA CONFORME ANEXO TECNICO,2019-06-18,2019-06-24,2019-09-30,No definido,DAP - Entregado en un punto (lugar de destino convenido),FANNY JANETH GUTIERREZ PRIETO,20484821,No,No,Mujer,CO,11970000.0,11970000,11970000,0,0,No Definido,Si,Distribuido,Funcionamiento,terminado,No,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.846657&isFromPublicArea=True&isModal=true&asPopupView=true'}
3,CO1.PCCNTR.1000908,AGENCIA LOGISTICA DE LAS FUERZAS MILITARES,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Licitación pública,V1.41115300,ADQUISICIÓN DE UN SISTEMA DE SONAR 3D DE ALTA RESOLUCIÓN CON DESTINO AL DEPARTAMENTO DE BUCEO Y SALVAMENTO,2019-07-25,2019-07-12,2019-11-30,Dia(s),Como acordado previamente,CASCO ANTIGUO,76044753,No,No,Otro,CO,733992000.0,0,0,733992000,0,No,No,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.826239&isFromPublicArea=True&isModal=true&asPopupView=true'}
4,CO1.PCCNTR.1000911,PARQUES NACIONALES NATURALES DE COLOMBIA - DIRECCION TERRITORIAL CARIBE,Bolívar,Cartagena,Territorial,Ambiente y Desarrollo Sostenible,Ejecutivo,Centralizada,Suministros,Mínima cuantía,V1.92121701,Suministro de equipos para la instalación de todas las cámaras necesarias para el monitoreo; para ser instalados en\nlas sedes oper

In [1549]:
""" 1.5 Muestra de 10 filas aleatorias del dataset. """
with pd.option_context('display.max_columns', None, 'display.max_colwidth', 20):
    display(secop_df.sample(n=10, random_state=42))

,id_contrato,nombre_entidad,departamento,ciudad,orden,sector,rama,entidad_centralizada,tipo_de_contrato,modalidad_de_contratacion,codigo_de_categoria_principal,objeto_del_contrato,fecha_de_firma,fecha_de_inicio_del_contrato,fecha_de_fin_del_contrato,duraci_n_del_contrato,condiciones_de_entrega,proveedor_adjudicado,documento_proveedor,es_pyme,es_grupo,g_nero_representante_legal,nacionalidad_representante_legal,valor_del_contrato,valor_pagado,valor_facturado,valor_pendiente_de_pago,dias_adicionados,habilita_pago_adelantado,el_contrato_puede_ser_prorrogado,origen_de_los_recursos,destino_gasto,estado_contrato,liquidaci_n,obligaci_n_ambiental,urlproceso
150471,CO1.PCCNTR.7348614,PERSONERIA MUNIC...,Valle del Cauca,No Definido,Territorial,No aplica/No per...,Corporación Autó...,Centralizada,Compraventa,Mínima cuantía,V1.84131500,ADQUISICIÓN DE L...,2025-01-29,2025-01-31,2026-01-31,364 Dia(s),No Definido,Aseguradora Soli...,860524654,No,No,Mujer,CO,8.613369e+06,0,0,8613369,0,No,No,Distribuido,Funcionamiento,En ejecución,Si,No,{'url': 'https:/...
103794,CO1.PCCNTR.5436333,SUBRED INTEGRADA...,Distrito Capital...,No Definido,Territorial,Salud y Protecci...,Corporación Autó...,Descentralizada,Compraventa,Contratación rég...,V1.42141800,ADQUISICIÓN DE E...,2023-10-10,2023-10-24,2024-02-23,4 Mes(es),Como acordado pr...,SMN SOLUCIONES I...,900343346,Si,No,No Definido,CO,5.862410e+07,58624100,58624100,1,0,No,No,Distribuido,Funcionamiento,terminado,No,No,{'url': 'https:/...
100976,CO1.PCCNTR.5359635,MUNICIPIO DE VEN...,Antioquia,Venecia,Territorial,No aplica/No per...,Ejecutivo,Centralizada,Suministros,Mínima cuantía,V1.31161503,Suministro de ma...,2023-09-14,2023-09-19,2023-10-03,15 Dia(s),Transporte incluido,DVCE INGENIERIA ...,901430321,Si,No,Hombre,CO,2.802001e+07,0,0,28020006,0,No,No,Distribuido,Inversión,terminado,No,No,{'url': 'https:/...
95393,CO1.PCCNTR.5200288,ALCALDÍA DEL DIS...,Bolívar,Cartagena,Territorial,No aplica/No per...,Ejecutivo,Centralizada,Compraventa,Selección abrevi...,V1.52141500,ADQUISICION E IN...,2023-07-21,2023-07-28,2023-09-27,60 Dia(s),Como acordado pr...,LICITACIONES Y A...,901058050,Si,No,No Definido,CO,5.415162e+08,0,0,541516228,0,No,No,Distribuido,Inversión,En ejecución,No,No,{'url': 'https:/...
3016,CO1.PCCNTR.1079143,ARMADA NACIONAL ...,Bolívar,Cartagena,Nacional,Salud y Protecci...,Ejecutivo,Descentralizada,Suministros,Mínima cuantía,V1.42271713,ADQUISICION; INS...,2019-08-29,2019-08-29,2019-11-15,Dia(s),No Definido,CASCO ANTIGUO,76044753,No,No,Otro,CO,2.676310e+07,0,0,26763100,26,No Definido,No,Distribuido,Inversión,Modificado,No,No,{'url': 'https:/...
194179,CO1.PCCNTR.946551,UNIDAD ADMINISTR...,Distrito Capital...,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Selección Abrevi...,V1.43233205,Contratar la ren...,2019-05-08,2019-05-03,2019-12-31,Dia(s),Como acordado pr...,Etek Internation...,8300615762,No,No,No Definido,CO,2.365117e+08,0,0,236511732,0,No Definido,Si,Distribuido,Funcionamiento,Cerrado,Si,No,{'url': 'https:/...
177103,CO1.PCCNTR.8419393,INSTITUTO DE DEP...,Antioquia,Medellín,Territorial,deportes,Ejecutivo,Centralizada,Suministros,Contratación dir...,V1.43211501,SUMINISTRAR E IN...,2025-10-06,2025-10-06,2026-02-20,137 Dia(s),No Definido,JUVALHER COLOMBI...,900703922,Si,No,Hombre,CO,2.588747e+08,0,0,258874683,64,No,No,Distribuido,Inversión,Modificado,No,No,{'url': 'https:/...
83396,CO1.PCCNTR.4667517,CONCEJO MUNICIPA...,Boyacá,Tunja,Territorial,Servicio Público,Corporación Autó...,Centralizada,Compraventa,Mínima cuantía,V1.84131601,SUMINISTRO DEL S...,2023-02-22,2023-02-23,2024-02-25,1 Año(s),No Definido,SEGUROS MUNDIAL ...,860037013,No,No,Hombre,CO,1.581000e+07,15810000,15810000,0,0,No,No,Distribuido,Funcionamiento,En ejecución,No,No,{'url': 'https:/...
103281,CO1.PCCNTR.5421145,FUERZA AEROESPAC...,Distrito Capital...,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Mínima cuantía,V1.25174810,ADQUISICIÓN DE H...,2023-10-05,2023-10-17,2023-10-31,27 Dia(s),Como acordado 

In [1550]:
""" 1.6 Deteccion y correccion de nombres de columna con problemas de encoding (tildes/enies perdidas). """

def fragmento_sospechoso(nombre_columna):
    fragmentos = nombre_columna.split('_')
    return any(len(frag) == 1 for frag in fragmentos)

columnas_sospechosas = [col for col in secop_df.columns if fragmento_sospechoso(col)]
print("Columnas con posible problema de encoding:", columnas_sospechosas)

correccion_columnas = {
    'duraci_n_del_contrato': 'duracion_del_contrato',
    'g_nero_representante_legal': 'genero_representante_legal',
    'liquidaci_n': 'liquidacion',
    'obligaci_n_ambiental': 'obligacion_ambiental',
}

secop_df = secop_df.rename(columns=correccion_columnas)
print("Columnas corregidas:", list(correccion_columnas.values()))

Columnas con posible problema de encoding: ['duraci_n_del_contrato', 'g_nero_representante_legal', 'liquidaci_n', 'obligaci_n_ambiental']
Columnas corregidas: ['duracion_del_contrato', 'genero_representante_legal', 'liquidacion', 'obligacion_ambiental']


### Aclaración de corrección de encoding: unica celda presente en Calidad de Datos que aplica correcciones, esto es debido a que el analisis lo requiere, este codigo esta repetido en eda_secop.
- ['duraci_n_del_contrato', 'g_nero_representante_legal', 'liquidaci_n', 'obligaci_n_ambiental']



## 2. Validez

In [1551]:
""" 2.1 Análisis de validez de secop_df """

print("Tipos de datos actuales:")
display(secop_df.dtypes)



Tipos de datos actuales:


id_contrato                                    str
nombre_entidad                                 str
departamento                                   str
ciudad                                         str
orden                                          str
sector                                         str
rama                                           str
entidad_centralizada                           str
tipo_de_contrato                               str
modalidad_de_contratacion                      str
codigo_de_categoria_principal                  str
objeto_del_contrato                            str
fecha_de_firma                      datetime64[us]
fecha_de_inicio_del_contrato        datetime64[us]
fecha_de_fin_del_contrato           datetime64[us]
duracion_del_contrato                          str
condiciones_de_entrega                         str
proveedor_adjudicado                           str
documento_proveedor                            str
es_pyme                        

#### Lista por tipo de variable 

In [1552]:
""" Cantidad de columnas por tipo de dato en pandas """

nombres_de_dtype = {'str': 'Texto', 'int64': 'Entero', 'float64': 'Float (decimal)', 'datetime64[us]': 'Fecha'}
resumen_de_dtypes = secop_df.dtypes.astype(str).value_counts().rename_axis('dtype_pandas').reset_index(name='Numero de columnas')
resumen_de_dtypes.insert(1, 'Tipo', resumen_de_dtypes['dtype_pandas'].map(nombres_de_dtype))

display(resumen_de_dtypes)

,dtype_pandas,Tipo,Numero de columnas
0,str,Texto,28
1,int64,Entero,4
2,datetime64[us],Fecha,3
3,float64,Float (decimal),1


#### Conteo por tipo de columnas
- El dataset cuenta 28 columnas tipo Texto, 4 tipo Entero, 3 tipo Fecha y 1 tipo Float

In [1553]:
""" Clasificación manual de variables por grupo y subgrupo"""

tabla_tipos = pd.DataFrame({
    'Grupo': ['Categórica', 'Categórica', 'Categórica',
              'Numérica', 'Numérica',
              'Temporal',
              'Otras', 'Otras', 'Otras', 'Otras'],
    'Subgrupo': ['Nominal (arbitrario)', 'Ordinal (ordenado)', 'Binaria (Si/No)',
                 'Continua (valores en dinero)', 'Discreta (numero de dias)',
                 'Fecha',
                 'Identificador', 'Texto libre', 'Código jerárquico (UNSPSC)', 'Mixta mal tipada'],
    'Variables': [
        'departamento, ciudad, orden, sector, rama, modalidad_de_contratacion, condiciones_de_entrega, '
        'nacionalidad_representante_legal, genero_representante_legal, destino_gasto, origen_de_los_recursos, '
        'tipo_de_contrato, entidad_centralizada',
        'estado_contrato',
        'es_pyme, es_grupo, el_contrato_puede_ser_prorrogado, liquidacion, obligacion_ambiental, habilita_pago_adelantado',
        'valor_del_contrato, valor_pagado, valor_facturado, valor_pendiente_de_pago',
        'dias_adicionados',
        'fecha_de_firma, fecha_de_inicio_del_contrato, fecha_de_fin_del_contrato',
        'id_contrato, documento_proveedor, urlproceso',
        'objeto_del_contrato, proveedor_adjudicado, nombre_entidad',
        'codigo_de_categoria_principal',
        'duracion_del_contrato',
    ],
    'Cantidad': [13, 1, 6, 4, 1, 3, 3, 3, 1, 1],
})


tabla_tipos = tabla_tipos.set_index(['Grupo', 'Subgrupo'])
pd.set_option('display.max_colwidth', None)
display(tabla_tipos)
print("Total de variables:", tabla_tipos['Cantidad'].sum())

Variables  \
Grupo      Subgrupo                                                                                                                                                                                                                                                                  
Categórica Nominal (arbitrario)          departamento, ciudad, orden, sector, rama, modalidad_de_contratacion, condiciones_de_entrega, nacionalidad_representante_legal, genero_representante_legal, destino_gasto, origen_de_los_recursos, tipo_de_contrato, entidad_centralizada   
           Ordinal (ordenado)                                                                                                                                                                                                                                      estado_contrato   
           Binaria (Si/No)                                                                                                                                        es_pyme, es_grupo, el_contrato_puede_ser_prorrogado, liquidacion, obligacion_ambiental, habilita_pago_adelantado   
Numérica   Continua (valores en dinero)                                                                                                                                                                 valor_del_contrato, valor_pagado, valor_facturado, valor_pendiente_de_pago   
           Discreta (numero de dias)                                                                                                                                                                                                                              dias_adicionados   
Temporal   Fecha                                                                                                                                                                                           fecha_de_firma, fecha_de_inicio_del_contrato, fecha_de_fin_del_contrato   
Otras      Identificador                                                                                                                                                                                                              id_contrato, documento_proveedor, urlproceso   
           Texto libre                                                                                                                                                                                                   objeto_del_contrato, proveedor_adjudicado, nombre_entidad   
           Código jerárquico (UNSPSC)                                                                                                                                                                                                                codigo_de_categoria_principal   
           Mixta mal tipada                                                                                                                                                                                                                                  duracion_del_contrato   

                                         Cantidad  
Grupo      Subgrupo                                
Categórica Nominal (arbitrario)                13  
           Ordinal (ordenado)                   1  
           Binaria (Si/No)                      6  
Numérica   Continua (valores en dinero)         4  
           Discreta (numero de dias)            1  
Temporal   Fecha                                3  
Otras      Identificador                        3  
           Texto libre                          3  
           Código jerárquico (UNSPSC)           1  
           Mixta mal tipada                     1

Total de variables: 36


### Resumen detallado de tipo de variables por categoria 

- Se evidencia:
    - 20 variables categoricas de las cuales, 6 son binarias, 13 son nominales, es decir no siguen un orden aparente y 1 de tipo ordinal dado que el si bien los tipos de estado de contrato no estan separados por una distancia cuantificable, si es posible concluir que en general un contrato sigue una serie de etapas:  primero se establece un borrador, luego entra en fase de ejecucion y posteriormente se marca como terminado, para finalmente identificarlo como cerrado, asi mismo puede ser suspendido, cancelado, cedido o modificado en cualquier etapa del mismo.
    - 5 Variables numericas, de estas 4 son continuas que corresponden a valor monetario de los contratos y 1 discreta que hace referencia a los dias adicionados.
    - 3 variables temporales de tipo fecha 
    - 8 variables que corresponden a 3 identificador unico, 3 de texto libre, 1 codigo jerarquico y 1 con topologia mal tipiada (observado en celda 1.5).


In [1554]:
""" 2.2 Conteo de tipo de fechas """

print(secop_df["fecha_de_firma"].map(type).value_counts(), "\n")
print(secop_df["fecha_de_inicio_del_contrato"].map(type).value_counts(), "\n")
print(secop_df["fecha_de_fin_del_contrato"].map(type).value_counts(), "\n")


fecha_de_firma
<class 'pandas.Timestamp'>    196391
Name: count, dtype: int64 

fecha_de_inicio_del_contrato
<class 'pandas.Timestamp'>             193951
<class 'pandas.api.typing.NaTType'>      2440
Name: count, dtype: int64 

fecha_de_fin_del_contrato
<class 'pandas.Timestamp'>    196391
Name: count, dtype: int64 



In [1555]:
""" 2.3 Rango de las fechas del contrato """

cols_fecha = ['fecha_de_firma', 'fecha_de_inicio_del_contrato', 'fecha_de_fin_del_contrato']

# Límites: el dataset cubre contratos firmados entre 2019 y 2025.
# Antes de 2015 no es plausible para estos contratos; después de 2030 implica un contrato de bienes de más de 5 años.
limite_inferior, limite_superior = pd.Timestamp('2015-01-01'), pd.Timestamp('2030-12-31')
fin_del_periodo = pd.Timestamp('2025-12-31')

filas = []
for col in cols_fecha:
    fecha = secop_df[col]
    filas.append({
        'columna': col,
        'nulos': fecha.isna().sum(),
        'minimo': fecha.min().date(),
        'maximo': fecha.max().date(),
        'antes_de_2015': (fecha < limite_inferior).sum(),
        'despues_de_2030': (fecha > limite_superior).sum(),
        'posteriores_a_2025 (vigentes)': (fecha > fin_del_periodo).sum(),
    })

rango_fechas = pd.DataFrame(filas).set_index('columna')
display(rango_fechas)

,nulos,minimo,maximo,antes_de_2015,despues_de_2030,posteriores_a_2025 (vigentes)
columna,,,,,,
fecha_de_firma,0,2019-01-01,2025-12-31,0,0,0
fecha_de_inicio_del_contrato,2440,2016-06-27,2026-12-16,0,0,378
fecha_de_fin_del_contrato,0,0202-12-31,2065-01-31,1,20,5604


In [1556]:
""" 2.4 Contrato con fecha de fin anterior a 2015 """

fin_antes_2015 = secop_df[secop_df['fecha_de_fin_del_contrato'] < limite_inferior]
print(f"Contratos con fecha de fin antes de 2015: {len(fin_antes_2015)}")

# Ver todas las columnas del contrato de arriba hacia abajo
with pd.option_context('display.max_rows', None, 'display.max_colwidth', 100):
    display(fin_antes_2015.T)

Contratos con fecha de fin antes de 2015: 1


,81109
id_contrato,CO1.PCCNTR.4351918
nombre_entidad,GOBERNACIÓN DEL CAUCA
departamento,Cauca
ciudad,Popayán
orden,Territorial
sector,Servicio Público
rama,Ejecutivo
entidad_centralizada,Centralizada
tipo_de_contrato,Suministros
modalidad_de_contratacion,Mínima cuantía


In [1557]:
""" 2.5 Calidad de los contratos con fecha de fin posterior a 2030: ¿la fecha coincide con la duración declarada? """

despues_2030 = secop_df[secop_df['fecha_de_fin_del_contrato'] > limite_superior].copy()

# Duración declarada en días (número x factor de la unidad)
partes_dur = despues_2030['duracion_del_contrato'].str.strip().str.extract(r'^(?P<numero>\d+)?\s*(?P<unidad>.*)$')
factor_a_dias = {'Hora(s)': 1/24, 'Dia(s)': 1, 'Semana(s)': 7, 'Mes(es)': 30, 'Año(s)': 365}
despues_2030['dias_declarados'] = pd.to_numeric(partes_dur['numero'], errors='coerce') * partes_dur['unidad'].map(factor_a_dias)

# Duración según las fechas (desde el inicio; si no hay fecha de inicio, desde la firma)
inicio = despues_2030['fecha_de_inicio_del_contrato'].fillna(despues_2030['fecha_de_firma'])
despues_2030['dias_segun_fechas'] = (despues_2030['fecha_de_fin_del_contrato'] - inicio).dt.days

# Coinciden (±10%): contrato de largo plazo real. No coinciden: la fecha de fin está mal digitada
despues_2030['cuadra_con_duracion'] = (
    (despues_2030['dias_declarados'] - despues_2030['dias_segun_fechas']).abs() <= 0.10 * despues_2030['dias_declarados']
)

print(f"Contratos con fecha de fin después de 2030: {len(despues_2030)}")
print(f"  - Cuadran con su duración declarada (contratos de largo plazo): {despues_2030['cuadra_con_duracion'].sum()}")
print(f"  - No cuadran (fecha de fin mal digitada): {(~despues_2030['cuadra_con_duracion']).sum()}")

# Detalle solo

Contratos con fecha de fin después de 2030: 20
  - Cuadran con su duración declarada (contratos de largo plazo): 10
  - No cuadran (fecha de fin mal digitada): 10


In [1558]:
""" 2.6 Valores negativos en columnas monetarias (violacion de dominio: un valor de contrato o pago no deberia ser negativo). """

cols_monetarias = ['valor_del_contrato', 'valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']
for col in cols_monetarias:
    negativos = (secop_df[col] < 0).sum()
    print(f"{col}: {negativos} filas negativas")

valor_del_contrato: 0 filas negativas
valor_pagado: 0 filas negativas
valor_facturado: 0 filas negativas
valor_pendiente_de_pago: 144 filas negativas


In [1559]:
""" 2.7 Análisis de duracion_del_contrato: resumen por unidad y validación por tipo de problema  """

duracion = secop_df['duracion_del_contrato'].str.strip()

# Separar número y unidad
partes = duracion.str.extract(r'^(?P<numero>\d+)?\s*(?P<unidad>.*)$')
numero = pd.to_numeric(partes['numero'], errors='coerce')

# Duración en días 
factor_a_dias = {'Hora(s)': 1/24, 'Dia(s)': 1, 'Semana(s)': 7, 'Mes(es)': 30, 'Año(s)': 365}
dias = numero * partes['unidad'].map(factor_a_dias)
dias_segun_fechas = (secop_df['fecha_de_fin_del_contrato'] - secop_df['fecha_de_inicio_del_contrato']).dt.days

# Resumen por unidad
resumen_unidades = (
    pd.DataFrame({'valor_original': duracion, 'numero': numero, 'unidad': partes['unidad']})
    .groupby('unidad', dropna=False)
    .agg(
        numero_contratos=('valor_original', 'size'),
        sin_numero=('numero', lambda s: s.isna().sum()),
        ejemplo_aleatorio=('valor_original', lambda s: s.sample(1, random_state=42).iloc[0]),
    )
    .sort_values('numero_contratos', ascending=False)
)
resumen_unidades['%'] = (resumen_unidades['numero_contratos'] / len(duracion) * 100).round(2)
display(resumen_unidades)

# Reglas de validación 
es_no_definido = duracion.isna() | (duracion.str.lower() == 'no definido')
condiciones = [
    es_no_definido.fillna(False),
    (~es_no_definido & numero.isna()).fillna(False),
    (numero == 0).fillna(False),
    # Número de 8 dígitos que se puede leer como fecha DDMMAAAA: se digitó una fecha en vez de una duración
    pd.to_datetime(partes['numero'].where(partes['numero'].str.len() == 8), format='%d%m%Y', errors='coerce').notna(),
    (dias > 3650).fillna(False),
    # Tolerancia de 31 días por la aproximación mes = 30 días
    ((dias - dias_segun_fechas).abs() > 31).fillna(False),
]
etiquetas = [
    '1. No definido',
    '2. Unidad sin número',
    '3. Duración cero',
    '4. Fecha digitada como duración',
    '5. Mayor a 10 años',
    '6. No cuadra con fecha de fin - fecha de inicio, mayor a 31 días',
]
problema_duracion = pd.Series(np.select(condiciones, etiquetas, default='OK'), index=secop_df.index)

# Resumen detallado por tipo de problema
resumen_problemas = (
    pd.DataFrame({'etiqueta': problema_duracion, 'valor_original': duracion})
    .groupby('etiqueta')
    .agg(
        numero_contratos=('valor_original', 'size'),
        ejemplo=('valor_original', lambda s: s.sample(1, random_state=42).iloc[0]),
    )
)
resumen_problemas['%'] = (resumen_problemas['numero_contratos'] / len(secop_df) * 100).round(2)
display(resumen_problemas)

,numero_contratos,sin_numero,ejemplo_aleatorio,%
unidad,,,,
Dia(s),124543,11327,145 Dia(s),63.42
Mes(es),61575,2,1 Mes(es),31.35
No definido,9490,9490,No definido,4.83
Año(s),520,0,3 Año(s),0.26
Semana(s),243,0,3 Semana(s),0.12
Hora(s),20,0,48 Hora(s),0.01


,numero_contratos,ejemplo,%
etiqueta,,,
1. No definido,9490,No definido,4.83
2. Unidad sin número,11329,Dia(s),5.77
3. Duración cero,5,0 Dia(s),0.00
4. Fecha digitada como duración,2,30092025 Mes(es),0.00
5. Mayor a 10 años,56,136 Mes(es),0.03
"6. No cuadra con fecha de fin - fecha de inicio, mayor a 31 días",21660,1 Mes(es),11.03
OK,153849,183 Dia(s),78.34


### Hallazgos de la duración: 

- 9.490 no definidos
- 11.329 sin número 
- 4 fechas digitadas 
- 56 de más de 10 años  
- 21.660 que no cuadran con las fechas

### Hallazgos

- El 100% de los registros de fecha_de_inicio_del_contrato, fecha_de_fin_del_contrato y fecha_de_firma tienen el mismo formato de fecha YYYY-MM-DD
- duracion_del_contrato presenta un error en formato, haciendo que tenga 6 problemas de calidad
- `valor_pendiente_de_pago` tiene 144 filas con valores negativos (las demas columnas monetarias no tienen negativos). Se investigo el patron en el notebook eda_secop.ipynb: el 100% de esos casos tiene `valor_pagado > valor_del_contrato`, es decir, es un sobrepago real y no un error de captura.

## 3. Completitud

In [1560]:
""" 3.1 Analisis de completitud de los datos de todas las columnas del DataFrame."""

total_instancias = len(secop_df)
print(f"Total de registros: {total_instancias:,}")


def analizar_completitud(df):
    # Nulos y strings vacíos/blancos, son tratados de igual manera como valores incompletos.
    filas = []
    for columna_test in secop_df.columns:
        serie = secop_df[columna_test]
        es_dato_nulo = serie.isnull()
        if serie.dtype == object:
            es_dato_nulo = es_dato_nulo | (serie.astype(str).str.strip() == "")
        nulos = int(es_dato_nulo.sum())
        no_nulos = total_instancias - nulos
        

        filas.append({
            "columna": columna_test,
            "Total Registros": total_instancias,
            "Valores Completos": no_nulos,
            "Valores Nulos o Vacios": nulos,
         
        })

    df_completitud = pd.DataFrame(filas)
    return df_completitud

df_completitud = analizar_completitud(secop_df)


print("\nRESUMEN DE COMPLETITUD")
display(df_completitud)



df_incompletas = df_completitud[df_completitud["Valores Nulos o Vacios"] >= 1].copy()
df_incompletas["% Nulos o Vacios"] = (
    df_incompletas["Valores Nulos o Vacios"] / total_instancias
) * 100


df_incompletas["% Nulos o Vacios"] = df_incompletas["% Nulos o Vacios"].round(4)

print("\nRESUMEN DE COLUMNAS CON NULOS O VACÍOS (>= 1)")
display(df_incompletas)






Total de registros: 196,391

RESUMEN DE COMPLETITUD


,columna,Total Registros,Valores Completos,Valores Nulos o Vacios
0,id_contrato,196391,196391,0
1,nombre_entidad,196391,196391,0
2,departamento,196391,196391,0
3,ciudad,196391,196391,0
4,orden,196391,196391,0
5,sector,196391,196391,0
6,rama,196391,196391,0
7,entidad_centralizada,196391,196391,0
8,tipo_de_contrato,196391,196391,0
9,modalidad_de_contratacion,196391,196391,0



RESUMEN DE COLUMNAS CON NULOS O VACÍOS (>= 1)


,columna,Total Registros,Valores Completos,Valores Nulos o Vacios,% Nulos o Vacios
13,fecha_de_inicio_del_contrato,196391,193951,2440,1.2424
18,documento_proveedor,196391,196387,4,0.0020


In [1561]:
"""3.2 Filtrar registros con documento_proveedor nulo o vacío. """
filtro = secop_df["documento_proveedor"].isna() | (secop_df["documento_proveedor"].astype(str).str.strip() == "")

# Mostrar la tabla completa con esos registros
registros_nulos = secop_df[filtro]

print(f"Se encontraron {len(registros_nulos):,} registros con documento_proveedor nulo o vacío")
display(registros_nulos)


Se encontraron 4 registros con documento_proveedor nulo o vacío


,id_contrato,nombre_entidad,departamento,ciudad,orden,sector,rama,entidad_centralizada,tipo_de_contrato,modalidad_de_contratacion,codigo_de_categoria_principal,objeto_del_contrato,fecha_de_firma,fecha_de_inicio_del_contrato,fecha_de_fin_del_contrato,duracion_del_contrato,condiciones_de_entrega,proveedor_adjudicado,documento_proveedor,es_pyme,es_grupo,genero_representante_legal,nacionalidad_representante_legal,valor_del_contrato,valor_pagado,valor_facturado,valor_pendiente_de_pago,dias_adicionados,habilita_pago_adelantado,el_contrato_puede_ser_prorrogado,origen_de_los_recursos,destino_gasto,estado_contrato,liquidacion,obligacion_ambiental,urlproceso
19847,CO1.PCCNTR.1802368,DEPARTAMENTO DE ANTIOQUIA//,Antioquia,Medellín,Territorial,Servicio Público,Ejecutivo,Centralizada,Suministros,Selección abreviada subasta inversa,V1.12352104,COMPRA DE RON 3 AÑOS DE AÑEJAMIENTO DE GRADO MINIMODE 75% V/V,2020-09-11,2020-09-11,2020-12-31,90 Dia(s),Como acordado previamente,UNION TEMPORAL RON PANAMA 2020,NaN,No,Si,No Definido,CO,6.374149e+10,0,0,63741494232,0,No Definido,No,Distribuido,Funcionamiento,terminado,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.1251670&isFromPublicArea=True&isModal=true&asPopupView=true'}
71874,CO1.PCCNTR.4117029,GOBERNACION DEL DEPARTAMENTO DEL CESAR,Cesar,Valledupar,Territorial,Servicio Público,Ejecutivo,Descentralizada,Suministros,Selección abreviada subasta inversa,V1.44103000,SUMINISTRO; INSTALACION; RECARGAS; REPUESTOS; REPARACION; MANTENIMIENTO PREVENTIVO Y CORRECTIVO DE LOS TONERES Y TINTAS ORIGINALES; EQUIPOS DE COMPUTO; EQUIPOS ACTIVOS DE RED; EQUIPOS DE OFICINA IMPRESORAS; FOTOCOPIADORAS; ESCANER; DATACENTER; EQUIPOS TECNOLOGICOS; UPS; SERVIDORES Y CABLEADO ESTRUCTURADO; PERTENECIENTES A LA GOBERNACION DEL DEPARTAMENTO DEL CESAR Y SUS DEPENDENCIAS,2022-10-14,2022-10-20,2022-12-30,78 Dia(s),No Definido,CONSORCIO TECNOSUMINISTROS & SERVICIOS DEL CESAR 2022,NaN,Si,Si,Hombre,CO,3.091562e+09,0,0,3091561509,0,No Definido,No,Distribuido,Inversión,En ejecución,Si,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.3315881&isFromPublicArea=True&isModal=true&asPopupView=true'}
109414,CO1.PCCNTR.5593886,CENTRAL ADMINISTRATIVA ESPECIALIZADA INGENIEROS CENAC INGENIEROS,Distrito Capital de Bogotá,Bogotá,Nacional,defensa,Ejecutivo,Centralizada,Compraventa,Licitación pública,V1.23153206,ADQUISICIÓN DEL ROBOT ANTIEXPLOSIVOS PLATAFORMA ROBÓTICA ANTIEXPLOSIVOS VEHICULOS TERRESTRES NO TRIPULADOS PARA RESPUESTA EOD No JEMPP CEDE10 DINCO COING CENAM ET 03085 ING 4,2023-12-14,2023-12-15,2024-04-20,25 Dia(s),NXTWY.DLVY.2,UNION TEMPORAL ROBOTEAM,NaN,No,Si,Hombre,CO,3.450500e+09,0,0,3450500000,0,No,No,Distribuido,Funcionamiento,terminado,No,No,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.5077917&isFromPublicArea=True&isModal=true&asPopupView=true'}
133958,CO1.PCCNTR.6704363,Rama Judicial – Dirección Seccional de Administración Judicial de Bucaramanga,Santander,Bucaramanga,Nacional,Ley de Justicia,Judicial,Centralizada,Compraventa,Selección abreviada subasta inversa,V1.44103103,Contratar en nombre de la Nación - Consejo Superior de Judicatura - Dirección Ejecutiva Seccional de Administración Judicial de Bucaramanga; la adquisición de consumibles de impresión para los Despachos Judiciales y Oficinas administrativas de la Seccional Santander; en las condiciones técnicas; de calidad y cantidades requeridas por la entidad,2024-09-16,2024-09-20,2024-10-15,30 Dia(s),NXTWY.DLVY.3,UT TONERS MAXPRINTOTALY-2024,NaN,Si,Si,Hombre,CO,4.145611e+08,0,0,414561126,0,No,Si,Distribuido,Funcionamiento,Cerrado,Si,Si,{'url': 'https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6548090&isFromPublicArea=True&isModal=true&asPopupView=true'}


In [1562]:
"""3.3 Top 5 proveedores con más contratos. """
total_registros = len(secop_df)

# Agrupar por proveedor y contar contratos
top_proveedores = (
    secop_df
    .groupby(["documento_proveedor", "proveedor_adjudicado"])
    .size()
    .reset_index(name="cantidad_contratos")
    .sort_values("cantidad_contratos", ascending=False)
)


# Mostrar Top 5
top_5 = top_proveedores.head(5)

print(f"Total de registros: {total_registros:,}")
print("\n📊 Top 5 proveedores con más contratos:")
display(top_5)



Total de registros: 196,391

📊 Top 5 proveedores con más contratos:


,documento_proveedor,proveedor_adjudicado,cantidad_contratos
9231,804003744,CENTRAL DE SUMINISTROS SAS,640
6605,59706955,Deicy Bravo Jojoa,579
11718,860002400,LA PREVISORA S.A. COMPAÑÍA DE SEGUROS,560
10345,828002423,DISTRIBUIDORA COLOMBIANA DE MEDICAMENTOS Y TECNOLOGIAS EN SALUD S.A.S. BIC - DISCOLMETS S.A.S. BIC,475
16756,900503041,SUPRISA S.A.S,471


In [1563]:
""" 3.4 Consolidado anual de valores pagados, facturados y pendientes de pago. """


# Agrupar por año
df_resumen_anual = (
    secop_df.groupby(secop_df['fecha_de_firma'].dt.year)[
        ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']
    ]
    .sum()
    .reset_index()
    .rename(columns={'fecha_de_firma': 'Año'})
)

# Función: expresar todo en billones de pesos para facilitar la lectura
def formatear_cifras_colombia(valor):
    if pd.isna(valor):
        return valor
    
    # Agrupar por billones de pesos en Colombia (12 ceros = 1 Billon)
    val = valor / 1e12
    sufijo = " B"
    
    # Formateo visual (Punto = miles, Coma = decimales)
    if val.is_integer():
        texto_num = f"{int(val):,}"
    else:
        texto_num = f"{val:,.2f}"
    
    texto_es = texto_num.replace(',', 'X').replace('.', ',').replace('X', '.')
    
    return f"{texto_es}{sufijo}"

# 4. Aplicar formato
for col in ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']:
    df_resumen_anual[col] = df_resumen_anual[col].apply(formatear_cifras_colombia)

display(df_resumen_anual)

,Año,valor_pagado,valor_facturado,valor_pendiente_de_pago
0,2019,"0,75 B","0,78 B","2,07 B"
1,2020,"2,32 B","2,45 B","4,29 B"
2,2021,"2,42 B","2,88 B","12.864,09 B"
3,2022,"3,41 B","4,60 B","4,60 B"
4,2023,"3,97 B","5,04 B","5,88 B"
5,2024,"5,40 B","6,36 B","7,22 B"
6,2025,"5,20 B","7,04 B","15,67 B"


In [1564]:
""" 3.5 Rastreo de un posible outlier, top 5 contratos del 2021 por valor pendiente de pago. """

# Filtro por contratos firmados en 2021
contratos_21 = secop_df[secop_df['fecha_de_firma'].dt.year == 2021]

# Top 5 por valor pendiente de pago
top5_pendiente_21 = contratos_21.nlargest(5, 'valor_pendiente_de_pago')

# Formato de cifras: B = billones, M = millones, misma convención de la celda anterior
def formatear_b_o_m(valor):
    if abs(valor) >= 1e12:
        texto, unidad = f"{valor / 1e12:,.2f}", "B"
    else:
        texto, unidad = f"{valor / 1e6:,.2f}", "M"
    # Formato de punto para miles, coma para decimales
    texto = texto.replace(',', 'X').replace('.', ',').replace('X', '.')
    return f"{texto} {unidad}"

# Copia solo para mostrar, los valores numéricos originales quedan intactos con convención de B o M según el numero de ceros.
top5_mostrar = top5_pendiente_21[['id_contrato', 'nombre_entidad', 'objeto_del_contrato']].copy()
top5_mostrar['valor_del_contrato'] = top5_pendiente_21['valor_del_contrato'].apply(formatear_b_o_m)
top5_mostrar['valor_pendiente_de_pago'] = top5_pendiente_21['valor_pendiente_de_pago'].apply(formatear_b_o_m)

display(top5_mostrar)

,id_contrato,nombre_entidad,objeto_del_contrato,valor_del_contrato,valor_pendiente_de_pago
50512,CO1.PCCNTR.3102623,INSTITUTO MUNICIPAL DEL DEPORTE Y LA RECREACION DE YUMBO,ADQUISICION A TITULO DE COMPRAVENTA DE TRES (3) MAQUINAS: DOS TRACTORES CORTA CESPED DE 23 CABALLOS Y UNA MAQUINA GRINERA MANUAL CORTACESPED AUTOPULSADA; PARA EL MANTENIMIENTO DE LA SUPERFICIE DE JUEGO EN GRAMA NATURAL DEL ESTADIO MUNICIPAL Y MANTENIMIENTO DE LAS DEMAS ZONAS VERDES EN OTROS ESCENARIOS DEPORTIVOS Y RECREATIVOS; A CARGO DEL IMDERTY; DE ACUERDO A LAS CONDICIONES TECNICAS QUE DE CADA MAQUINA DE DESCRIBEN Y DETALLAN EN LAS ESPECIFICACIONES TECNICAS.,"12.858,45 B","12.858,45 B"
50697,CO1.PCCNTR.3105869,EJERCITO DIRECCION DE ADQUISICIONES,EL CONTRATISTA SE OBLIGA CON EL M. D. N. - EJÉRCITO NACIONAL; AL ABASTECIMIENTO DE ESTANCIAS DE ALIMENTACIÓN CON DESTINO AL PERSONAL DE SOLDADOS Y ALUMNOS DE LAS UNIDADES MILITARES DEL EJÉRCITO NACIONAL; de acuerdo anexo No 1 de este contrato; el anexo No. 2 de la solicitud de oferta del proceso y la oferta presentada por el contratista; que hacen parte integral del presente contrato.,"496.367,03 M","496.367,03 M"
39502,CO1.PCCNTR.2738469,EJERCITO DIRECCION DE ADQUISICIONES,EL CONTRATISTA SE OBLIGA CON EL M. D. N. - EJÉRCITO NACIONAL; AL ABASTECIMIENTO DE ESTANCIAS DE ALIMENTACIÓN CON DESTINO AL PERSONAL DE SOLDADOS Y ALUMNOS DE LAS UNIDADES MILITARES DEL EJÉRCITO NACIONAL; de acuerdo anexo No 1 de este contrato; el anexo No. 2 de la solicitud de oferta del proceso y la oferta presentada por el contratista,"186.507,07 M","186.507,07 M"
43630,CO1.PCCNTR.2891745,ALCALDIA MUNICIPIO DE APARTADO,COMPRA DE MOBILARIO REQUERIDO PARA LAS DIFERENTES DEPENDENCIAS DE LA ADMINISTRACION MUNICIPAL,"96.990,00 M","96.893,01 M"
7399,CO1.PCCNTR.1195435,INVIAS,SUMINISTRO DE MEZCLA ASFÁLTICA VÍA 4015,"81.995,06 M","81.995,06 M"


In [1565]:
""" 3.6 Diferencia entre el primero y el segundo contrato con mayor valor pendiente """

# Valor pendiente del primero y del segundo 
primero = top5_pendiente_21['valor_pendiente_de_pago'].iloc[0]
segundo = top5_pendiente_21['valor_pendiente_de_pago'].iloc[1]

# Cuántas veces es más grande el primero con respecto al segundo
veces = primero / segundo

# Cuanto es mas grande el primero que el segundo, expresado como diferencia porcentual
diferencia_pct = (primero - segundo) / segundo * 100

# Proporción que ocupa el primero respecto al total del año 2021
participacion = primero / contratos_21['valor_pendiente_de_pago'].sum() * 100


print(f"El primero es {veces:,.1f} veces el segundo")
print(f"Diferencia porcentual: {diferencia_pct:,.1f}% más grande")
print(f"Ese contrato corresponde al {participacion:.2f}% de todo el valor pendiente de pago del año 2021")

El primero es 25,905.1 veces el segundo
Diferencia porcentual: 2,590,412.6% más grande
Ese contrato corresponde al 99.96% de todo el valor pendiente de pago del año 2021


### Resumen de la diferencia entre el primer y segundo contrato con mayor valor pendiente
- Se evidencia:
    - El valor del primero es casi 26 mil veces el segundo
    - La diferencia porcentual es de 2,5 millones % mas grande con respecto al segundo 
    - el solo contrato por los 3 tractores corta cesped ocupa el 99.96% de la totalidad del valor pendiente de pago del año 2021

Se concluye que es un valor anomalo que probablemente fue mal digitado 

In [1566]:
"""3.7 Análisis de formato de las columnas financieras: valor_pagado, valor_facturado y valor_pendiente_de_pago. """

cols_financieras = ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']

# Función para analizar qué caracteres especiales contiene cada columna
def auditar_formato_texto(df, columnas):
    diagnostico = []
    for col in columnas:
        serie_str = df[col].astype(str).str.strip()
        
        tiene_punto = serie_str.str.contains(r'\.', regex=True).sum()
        tiene_coma = serie_str.str.contains(r',', regex=True).sum()
        tiene_ambos = serie_str.str.contains(r'\.', regex=True) & serie_str.str.contains(r',', regex=True)
        solo_numeros = serie_str.str.match(r'^\d+$').sum()
        
        diagnostico.append({
            "Columna": col,
            "Tiene Punto (.)": tiene_punto,
            "Tiene Coma (,)": tiene_coma,
            "Tiene Ambos (. y ,)": tiene_ambos.sum(),
            "Solo Dígitos (Sin separadores)": solo_numeros,
            "Muestras de Texto": serie_str.unique()[:3].tolist() # Ver 3 ejemplos reales
        })
    return pd.DataFrame(diagnostico)

display(auditar_formato_texto(secop_df, cols_financieras))


,Columna,Tiene Punto (.),"Tiene Coma (,)","Tiene Ambos (. y ,)",Solo Dígitos (Sin separadores),Muestras de Texto
0,valor_pagado,0,0,0,196391,"[17896237, 389999685, 11970000]"
1,valor_facturado,0,0,0,196391,"[17896237, 389999685, 11970000]"
2,valor_pendiente_de_pago,0,0,0,196247,"[1679, 315, 0]"


### Decisiones

- Existen 2,440 Nulos en fecha_de_inicio_del_contrato y 4 documento_proveedor: se dejan como valores faltantes reales (NaN), sin imputar. 
- Anomalia de 2021, el valor_pendiente_de_pago equivale a 41 veces el Presupuesto General de la Nacion de ese año:  se procede a la eliminacion del contrato CO1.PCCNTR.3102623 en eda_secop

In [1567]:
"""3.8 Asimetria y Curtosis de las columnas financieras: valor_pagado, valor_facturado y valor_pendiente_de_pago. """
columnas_evaluadas = ['valor_pagado', 'valor_facturado', 'valor_pendiente_de_pago']

# 1. Cálculo sobre la totalidad de los datos
resumen_estadistico = []

for col in columnas_evaluadas:
    skew_val = secop_df[col].skew()
    kurt_val = secop_df[col].kurt()  # Nota: Pandas calcula la curtosis de exceso (normal = 0)
    
    resumen_estadistico.append({
        "Columna": col,
        "Asimetría (Skewness)": round(skew_val, 4),
        "Curtosis (Kurtosis)": round(kurt_val, 4)
    })

df_métricas = pd.DataFrame(resumen_estadistico)

print("MÉTRICAS SOBRE DATOS COMPLETOS (INCLUYENDO ATÍPICOS)")
display(df_métricas)


MÉTRICAS SOBRE DATOS COMPLETOS (INCLUYENDO ATÍPICOS)


,Columna,Asimetría (Skewness),Curtosis (Kurtosis)
0,valor_pagado,168.3456,36616.2554
1,valor_facturado,137.6966,23803.6522
2,valor_pendiente_de_pago,443.1602,196390.9506


### Hallazgos

- fecha_de_inicio_del_contrato presenta 2440 registros nulos lo cual es el 1.24 % del total de este atributo. 
- Se encontraron 4 registros con documento_proveedor nulo o vacío
- en 2021 el valor pendiente de pago asciende a 13 mil billones de pesos sin embargo el Presupuesto General de la nación aprobado para ese año fue de $313,9 billones de pesos, lo que traduze en 41 veces el presupuesto de todo un pais, el valor pagado mas el valor facturado es un total de 5.3 Billones.
- La presencia del outlier en valor_pendiente_pago del 2021 da como resultado una curtosis extrema (tipo leptocurtica).

## 4. Unicidad

In [1568]:
""" 4.1 Análisis de unicidad de de id_contrato contra fecha_de_firma, fecha_de_inicio_del_contrato y fecha_de_fin_del_contrato """

total_registros = len(secop_df)
print(f"Total de registros: {total_registros:,}")

# Ajustar a la(s) columna(s) que deberían identificar un registro de forma única
columnas_clave = [
    "id_contrato", "fecha_de_firma", "fecha_de_inicio_del_contrato", "fecha_de_fin_del_contrato"
]


duplicados = (
    secop_df
    .groupby(columnas_clave)
    .size()
    .reset_index(name="cantidad_duplicados")
    .query("cantidad_duplicados > 1")
    .sort_values("cantidad_duplicados", ascending=False)
)

num_duplicados = len(duplicados)
if num_duplicados > 0:
    print(f"Se encontraron {num_duplicados:,} llaves duplicadas:")
    display(duplicados.head(20))

    print("\nEjemplos de registros duplicados:")
    primera_llave = duplicados.iloc[0][columnas_clave]
    filtro = (secop_df[columnas_clave] == primera_llave.values).all(axis=1)
    display(secop_df[filtro])
else:
    columnas_str = " + ".join(columnas_clave)
    print(f"No se encontraron duplicados por la combinacion de  {columnas_str}")
    


Total de registros: 196,391
No se encontraron duplicados por la combinacion de  id_contrato + fecha_de_firma + fecha_de_inicio_del_contrato + fecha_de_fin_del_contrato


In [1569]:
""" 4.2 Analisis de duplicados de la combinacion de fecha_de_inicio_del_contrato + id_contrato + proveedor_adjudicado. """

columnas_clave = ["fecha_de_inicio_del_contrato", "id_contrato", "proveedor_adjudicado"]

# Contar duplicados
duplicados = (
    secop_df
    .groupby(columnas_clave)
    .size()
    .reset_index(name="cantidad_duplicados")
    .query("cantidad_duplicados > 1")
    .sort_values("cantidad_duplicados", ascending=False)
)

# Resultado
num_duplicados = len(duplicados)
if num_duplicados > 0:
    print(f"Se encontraron {num_duplicados:,} combinaciones duplicadas de {columnas_clave}:")
    display(duplicados.head(20))

    print("\nEjemplo de registros duplicados:")
    primera_llave = duplicados.iloc[0][columnas_clave]
    filtro = (secop_df[columnas_clave] == primera_llave.values).all(axis=1)
    display(secop_df[filtro])
else:
    columnas_str = " + ".join(columnas_clave)
    print(f"La combinación de {columnas_str} es única en todos los registros")

La combinación de fecha_de_inicio_del_contrato + id_contrato + proveedor_adjudicado es única en todos los registros


### Decisiones
- Ambas combinaciones de llave probadas (id_contrato + fechas; fecha_de_inicio + id_contrato + proveedor_adjudicado) resultaron unicas, sin duplicados. Se confirma que la granularidad del dataset es una fila = un `id_contrato`. No se requiere ningun tratamiento de deduplicacion.

## 5. Consistencia

In [1570]:
""" 5.1 Analisis de consistencia de fechas dentro del mismo registro (violacion de integridad a nivel de registro). """

firma_posterior_a_fin = (secop_df["fecha_de_firma"] > secop_df["fecha_de_fin_del_contrato"]).sum()
firma_posterior_a_inicio = (secop_df["fecha_de_firma"] > secop_df["fecha_de_inicio_del_contrato"]).sum()
fin_anterior_a_inicio = (secop_df["fecha_de_fin_del_contrato"] < secop_df["fecha_de_inicio_del_contrato"]).sum()

print(f"fecha_de_firma posterior a fecha_de_fin_del_contrato: {firma_posterior_a_fin:,} registros")
print(f"fecha_de_firma posterior a fecha_de_inicio_del_contrato: {firma_posterior_a_inicio:,} registros")
print(f"fecha_de_fin_del_contrato anterior a fecha_de_inicio_del_contrato: {fin_anterior_a_inicio:,} registros")

fecha_de_firma posterior a fecha_de_fin_del_contrato: 2,601 registros
fecha_de_firma posterior a fecha_de_inicio_del_contrato: 13,047 registros
fecha_de_fin_del_contrato anterior a fecha_de_inicio_del_contrato: 125 registros


### Decisiones
- Se encontraron 2,601 registros con `fecha_de_firma` posterior a `fecha_de_fin_del_contrato`, 13,047 con `fecha_de_firma` posterior a `fecha_de_inicio_del_contrato`, y 125 con `fecha_de_fin_del_contrato` anterior a `fecha_de_inicio_del_contrato`. Son secuencias temporales logicamente imposibles.
- No se eliminan filas ni se corrigen las fechas (no hay forma de saber cual es el valor correcto). En el notebook `eda_secop.ipynb` estas inconsistencias, junto con las fechas fuera de rango (2016-2025), se marcaron con una columna booleana `fecha_sospechosa` para poder excluirlas puntualmente de analisis de duracion/plazo sin perder el resto de la informacion del contrato. Se recomienda aplicar el mismo criterio aqui para mantener consistencia entre los dos notebooks.